File: backend/node_version/src/controllers/authController.ts
// controllers/authController.ts
import { Request, Response } from 'express';
import { handleOAuthLogin } from '../services/authService';
import { verifyGoogleToken, verifyMicrosoftToken } from '../services/tokenService'; 

// Google login route
export const googleLogin = async (req: Request, res: Response) => {
  const { token } = req.body; // Extract the token from the request body
  try {
    const { email, name } = await verifyGoogleToken(token); // Verify token and extract user info

    if(email == null || name == null){
      res.status(400).json({ message: "Please sign up with Google" });
    }

    const user = await handleOAuthLogin(email, name);
    // Initialize session user data
    req.session.user = {
      id: user.id,
      email: user.email,
      name: user.name,
      role: user.role,
    };

    res.json(user);
  } catch (error: any) {
    res.status(500).json({ message: error.message });
  }
};

// Microsoft login route
export const microsoftLogin = async (req: Request, res: Response) => {
  const { token } = req.body; // Extract the token from the request body

  try {
    const { email, name } = await verifyMicrosoftToken(token); // Verify token and extract user info

    if(email == null || name == null){
      res.status(400).json({ message: "Please sign up with Microsoft" });
    }

    const user = await handleOAuthLogin(email, name);

    // Initialize session user data
    req.session.user = {
      id: user.id,
      email: user.email,
      name: user.name,
      role: user.role, 
    };
    res.json(user);
  } catch (error: any) {
    res.status(500).json({ message: error.message });
  }
};

export const logout = (req: Request, res: Response) => {
  // Clear the session and send a response
  req.session.destroy((err) => {
      if (err) {
          return res.status(500).json({ message: 'Logout failed' });
      }
      res.clearCookie('connect.sid'); // Clear the session cookie
      res.status(200).json({ message: 'Logged out successfully' });
  });
};

File: backend/scala-seed-project/src/controllers/AuthController.scala
// src/controllers/AuthController.scala

import cats.effect.Sync
import org.http4s.{Request, Response}
import org.http4s.dsl.Http4sDsl
import org.http4s.HttpRoutes
import io.circe.generic.auto._
import io.circe.syntax._
import services.{AuthService, TokenService}

case class LoginRequest(token: String)

object AuthController {
  def routes[F[_]: Sync](authService: AuthService[F], tokenService: TokenService[F]): HttpRoutes[F] = {
    val dsl = new Http4sDsl[F]{}
    import dsl._

    HttpRoutes.of[F] {
      // Google login route
      case req @ POST -> Root / "auth" / "google" =>
        for {
          loginRequest <- req.as[LoginRequest]
          userInfo <- tokenService.verifyGoogleToken(loginRequest.token).handleError(_ => None)
          response <- userInfo match {
            case Some((email, name)) =>
              for {
                user <- authService.handleOAuthLogin(email, name)
                _ = req.attributes.insert(UserSessionAttr, user) // Assume UserSessionAttr stores session info
                resp <- Ok(user.asJson)
              } yield resp
            case None => BadRequest("Please sign up with Google")
          }
        } yield response

      // Microsoft login route
      case req @ POST -> Root / "auth" / "microsoft" =>
        for {
          loginRequest <- req.as[LoginRequest]
          userInfo <- tokenService.verifyMicrosoftToken(loginRequest.token).handleError(_ => None)
          response <- userInfo match {
            case Some((email, name)) =>
              for {
                user <- authService.handleOAuthLogin(email, name)
                _ = req.attributes.insert(UserSessionAttr, user) // Assume UserSessionAttr stores session info
                resp <- Ok(user.asJson)
              } yield resp
            case None => BadRequest("Please sign up with Microsoft")
          }
        } yield response

      // Logout route
      case DELETE -> Root / "auth" / "logout" =>
        req.attributes.remove(UserSessionAttr) // Clear the session data
        Ok("Logged out successfully")
    }
  }
}
